### PR TITLE
[#171211343] Add user data processing API

### PR DESCRIPTION
**Pull request by AleDore**
Exposing APIs for user data processing management

### CODE REVIEW

**Comment by digitalcitizenship** [comment]
## Affected stories

  * 🌟 [#171211343](https://www.pivotaltracker.com/story/show/171211343): Raccogliere la volontà dell'utente di veder cancellato il proprio account e avere a disposizione l'esportazione dei propri dati


New dependencies added: <a href='https://www.npmjs.com/package/debug'><code>debug</code></a>.


<h2><a href="https://github.com/visionmedia/debug#readme">debug</a></h2>
<p>Author: TJ Holowaychuk</p>
<p>Description: small debugging utility</p>
<p>Homepage: <a href="https://github.com/visionmedia/debug#readme">https://github.com/visionmedia/debug#readme</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>about 8 years ago</td></tr><tr><td>Last Updated</td><td>3 months ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>4</td></tr><tr><td>Releases</td><td>64</td></tr><tr><td>Direct Dependencies</td><td><a href='https://www.npmjs.com/package/ms'><code>ms</code></a></td></tr><tr><td>Keywords</td><td>debug, log and debugger</td></tr>
</table>
This README is too long to show.

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: utils/middlewares/userDataProcessing.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import {
+  NewUserDataProcessing,
+  UserDataProcessing
+} from "io-functions-commons/dist/src/models/user_data_processing";
+import { IRequestMiddleware } from "io-functions-commons/dist/src/utils/request_middleware";
+
+/**
+ * A middleware that extracts a NewUserDataProcessing payload from a request.
+ */
+export const NewUserDataProcessingMiddleware: IRequestMiddleware<
```
Comment (gunzip) [inline utils/middlewares/userDataProcessing.ts]: I've added a middleware that you can use inline to parse body payload:
https://github.com/pagopa/io-functions-commons/blob/master/src/utils/middlewares/required_body_payload.ts

can you try using this (1.12.0 version of io-functions-commons package) and get rid of this new middlewares here ?

Comment (AleDore) [reply utils/middlewares/userDataProcessing.ts]: Do you mean refactoring this way?
```
import { UserDataProcessingChoiceRequest } from "io-functions-commons/dist/generated/definitions/UserDataProcessingChoiceRequest";
import {
  NewUserDataProcessing,
  UserDataProcessing
} from "io-functions-commons/dist/src/models/user_data_processing";
import { RequiredBodyPayloadMiddleware } from "io-functions-commons/dist/src/utils/middlewares/required_body_payload";
/**
 * A middleware that extracts a NewUserDataProcessing payload from a request.
 */
export const NewUserDataProcessingMiddleware = RequiredBodyPayloadMiddleware(
  NewUserDataProcessing
);

/**
 * A middleware that extracts a UserDataProcessingChoice payload from a request.
 */
export const UserDataProcessingChoiceMiddleware = RequiredBodyPayloadMiddleware(
  UserDataProcessingChoiceRequest
);

/**
 * A middleware that extracts a UserDataProcessing payload from a request.
 */
export const UserDataProcessingMiddleware = RequiredBodyPayloadMiddleware(
  UserDataProcessing
);
```

Comment (gunzip) [reply utils/middlewares/userDataProcessing.ts]: no I mean use it inline in the handler.ts without assign new names
(the usage is right anyway : )

File: CreateUserDataProcessing/handler.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+import { UserDataProcessingStatusEnum } from "io-functions-commons/dist/generated/definitions/UserDataProcessingStatus";
+import {
+  makeUserDataProcessingId,
+  UserDataProcessing,
+  UserDataProcessingModel
+} from "io-functions-commons/dist/src/models/user_data_processing";
+import { UserDataProcessingChoiceMiddleware } from "../utils/middlewares/userDataProcessing";
+
+/**
+ * Type of an CreateProfile handler.
```
Comment (gunzip) [inline CreateUserDataProcessing/handler.ts]:
```suggestion
 * Type of an CreateUserDataProcessing handler.
```

File: CreateUserDataProcessing/handler.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  | IResponseErrorValidation
+  | IResponseErrorQuery
+  | IResponseErrorConflict
+>;
+
+export function CreateUserDataProcessingHandler(
+  userDataProcessingModel: UserDataProcessingModel
+): ICreateUserDataProcessingHandler {
+  return async (context, fiscalCode, createUserDataProcessingPayload) => {
+    const logPrefix = `CreateUserDataProcessingHandler|FISCAL_CODE=${fiscalCode}`;
```
Comment (gunzip) [inline CreateUserDataProcessing/handler.ts]: can we just log the first 6 chars of fiscalcode ?

File: utils/middlewares/userDataProcessing.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import { UserDataProcessingChoiceRequest } from "io-functions-commons/dist/generated/definitions/UserDataProcessingChoiceRequest";
+import {
+  NewUserDataProcessing,
+  UserDataProcessing
+} from "io-functions-commons/dist/src/models/user_data_processing";
+import { RequiredBodyPayloadMiddleware } from "io-functions-commons/dist/src/utils/middlewares/required_body_payload";
+/**
+ * A middleware that extracts a NewUserDataProcessing payload from a request.
+ */
+export const NewUserDataProcessingMiddleware = RequiredBodyPayloadMiddleware(
```
Comment (gunzip) [inline utils/middlewares/userDataProcessing.ts]: can we use this inline in handler.ts ?

File: CreateUserDataProcessing/handler.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  userDataProcessingChoiceRequest: UserDataProcessingChoiceRequest
+) => Promise<
+  // tslint:disable-next-line: max-union-size
+  | IResponseSuccessJson<UserDataProcessing>
+  | IResponseErrorValidation
+  | IResponseErrorQuery
+  | IResponseErrorConflict
+>;
+
+export function CreateUserDataProcessingHandler(
```
Comment (gunzip) [inline CreateUserDataProcessing/handler.ts]: since we're calling createOrUpdateByNewOne this should be:

```suggestion
export function UpsertUserDataProcessingHandler(
```

(same for file names)

File: CreateUserDataProcessing/handler.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+      const createdOrUpdatedUserDataProcessing =
+        errorOrCreatedUserDataProcessing.value;
+      return ResponseSuccessJson(createdOrUpdatedUserDataProcessing);
+    }
+  };
+}
+
+/**
+ * Wraps an CreateProfile handler inside an Express request handler.
```
Comment (gunzip) [inline CreateUserDataProcessing/handler.ts]:
```suggestion
 * Wraps an CreateUserDataProcessing handler inside an Express request handler.
```

File: GetUserDataProcessing/__tests__/handler.test.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    expect(
+      userDataProcessingModelMock.findOneUserDataProcessingById
+    ).toHaveBeenCalledWith(aFiscalCode, aUserDataProcessingId);
+    expect(response.kind).toBe("IResponseSuccessJson");
+    if (response.kind === "IResponseSuccessJson") {
+      expect(response.value).toEqual(aUserDataProcessing);
+    }
+  });
+
+  it("should respond with NotFound if profile does not exist", async () => {
```
Comment (gunzip) [inline GetUserDataProcessing/__tests__/handler.test.ts]:
```suggestion
  it("should respond with NotFound if userDataProcessing does not exist", async () => {
```

File: utils/middlewares/userDataProcessing.ts (hunk 4 above)
Comment (AleDore) [reply utils/middlewares/userDataProcessing.ts]: I suppose that we can't, because in request payload we have the object wrapper of a UserDataProcessingChoice ( we use UserDataProcessingChoiceRequestMiddleware for this purpose ).

File: package.json [hunk 8]
```diff
@@ -25,6 +25,7 @@
     "@types/nodemailer": "^4.6.8",
     "@types/superagent": "^4.1.3",
     "azurite": "^3.1.2-preview",
+    "debug": "^4.1.1",
```
Comment (gunzip) [inline package.json]: why is this needed ?

**Comment by gunzip** [comment]
congrats :+1:
